File: 0001-traversing-a-matrix/cpp/main.cpp
#include <iostream>
#include <chrono>

#define MATRIX_SIZE 20000

typedef unsigned int matrix_t;

int main() {
    auto matNumbers = new matrix_t[MATRIX_SIZE][MATRIX_SIZE];

    for(int i = 0; i < MATRIX_SIZE; i++){
        for(int j = 0; j < MATRIX_SIZE; j++){
            matNumbers[i][j] = random();
        }
    }

    float result = 0;

    clock_t begin = std::clock();
    for(int i = 0; i < MATRIX_SIZE; i++){
        for(int j = 0; j < MATRIX_SIZE; j++){
            result += matNumbers[i][j];
        }
    }

    clock_t end = std::clock();

    const double elapsed_time_row_mayor = ((double)(end - begin)) / CLOCKS_PER_SEC;

    printf("Sum: %f\n", result);

    result = 0;

    begin = std::clock();
    for(int j = 0; j < MATRIX_SIZE; j++){
        for(int i = 0; i < MATRIX_SIZE; i++){
            result += matNumbers[i][j];
        }
    }

    end = std::clock();

    const double elapsed_time_col_mayor = ((double)(end - begin)) / CLOCKS_PER_SEC;

    printf("Sum: %f\n", result);
    printf("elapsed time relation (col_mayor / row mayor) %f seconds\n", elapsed_time_col_mayor / elapsed_time_row_mayor);

    return 0;
}

File: 0001-traversing-a-matrix/javascript/main.js
const MATRIX_SIZE = 10000;
let arr = [];

for(let i = 0; i < MATRIX_SIZE; i++){
    arr[i] = [];
    for(let j = 0; j < MATRIX_SIZE; j++){
        arr[i][j] = Math.random();
    }
}

let result = 0;

let beginTime = new Date();
for(let j = 0; j < MATRIX_SIZE; j++){
    for(let i = 0; i < MATRIX_SIZE; i++){
        result += arr[i][j] ;
    }
}
let endTime_column = new Date() - beginTime;

console.log('time by cache UNFriendly: ', endTime_column);
console.log('result ' + result);

beginTime = new Date();
for(let j = 0; j < MATRIX_SIZE; j++){
    for(let i = 0; i < MATRIX_SIZE; i++){
        result += arr[j][i] ;
    }
}
let endTime_row = new Date() - beginTime;

console.log('time by cache friendly: ', endTime_row);

console.log('time difference: column / row', endTime_column / endTime_row);

